src/v/cluster/members_backend.cc

```diff
@@ -569,107 +629,146 @@ void members_backend::calculate_reallocations_batch(
+            // do not try to move internal partitions
+            if (
+              tp_ns.ns == model::kafka_internal_namespace
+              || tp_ns.ns == model::redpanda_ns) {
                 continue;
             }
-
-            if (rand_idx % sizeof(uint64_t) == 0) {
-                sampling_bytes = random_generators::get_int(
-                  std::numeric_limits<uint64_t>::max());
-                rand_idx = 0;
-            }
-
-            idx_it->second++;
-
-            if (sampling_bytes.test(rand_idx++)) {
+            if (!metadata.is_topic_replicable()) {
                 continue;
             }
+            // do not move topics that were created after node was added, they
+            // are allocated with new cluster capacity
+            if (meta.update) {
+                if (metadata.get_revision() > meta.update->offset()) {
+                    vlog(
+                      clusterlog.debug,
+                      "skipping reallocating topic {}, its revision {} is "
+                      "greater than node update {}",
+                      tp_ns,
+                      metadata.get_revision(),
+                      meta.update->offset);
+                    continue;
+                }
+            }
 
-            std::erase_if(to_move_from_node, [](const replicas_to_move& v) {
-                return v.left_to_move == 0;
-            });
-
-            std::sort(to_move_from_node.begin(), to_move_from_node.end(), cmp);
-            for (auto& to_move : to_move_from_node) {
-                if (
-                  is_in_replica_set(p.replicas, to_move.id)
-                  && to_move.left_to_move > 0) {
+            for (const auto& p : metadata.get_assignments()) {
+                if (is_in_replica_set(p.replicas, to_move.id)) {
                     partition_reallocation reallocation(
                       model::ntp(tp_ns.ns, tp_ns.tp, p.id), p.replicas.size());
                     reallocation.replicas_to_remove.emplace(to_move.id);
                     auto current_assignment = p;
-                    reassign_replicas(current_assignment, reallocation);
-                    // skip if partition was reassigned to the same node
-                    if (is_reassigned_to_node(reallocation, to_move.id)) {
+                    reassign_replicas(
+                      allocator, current_assignment, reallocation);
+
+                    if (!reallocation.allocation_units.has_value()) {
+                        continue;
+                    }
+                    const auto& new_assignment = reallocation.allocation_units
+                                                   .value()
+                                                   .get_assignments()
+                                                   .begin()
+                                                   ->replicas;
+
+                    auto new_error = error_after_move(
+                      total_replicas,
+                      node_replicas,
```

Comment:
hmm, should we update node_replicas if we use it to calculate the error after several moves?

Reply:
This is problematic as after we have already a batch of potential reassignments selected then we want to replace some of them. At this point a selected reassignment will not improve an error as that is the error is already minimized by the other reassignments that are there in the batch. On the other hand we do not want movements that are not improving the initial unevenness f.e. moving a partition between the nodes that weren't added or removed.

Reply:
> This is problematic as after we have already a batch of potential reassignments selected then we want to replace some of them. At this point a selected reassignment will not improve an error as that is the error is already minimized by the other reassignments that are there in the batch.

we can solve this by calculating the error delta that we'll get after replacing one reassignment with another